modules/http/src/main/java/org/mule/module/http/internal/listener/HttpResponseBuilder.java

```diff
@@ -209,7 +211,11 @@ else if (!configuredContentType.startsWith(HttpHeaders.Values.APPLICATION_X_WWW_
                             setupChunkedEncoding(httpResponseHeaderBuilder);
                         }
                         if (isStream) {
-                            httpEntity = new InputStreamHttpEntity((InputStream) payload);
+                            try {
+                                httpEntity = new InputStreamHttpEntity(IOUtils.toInputStream(payload.toString()));
```

Comment:
I don't think this is a good idea. Yes, it probably fixes the issue but it defeats the purpose of having a stream.
Here we are reading the whole contents of the stream and storing them in memory (inside the `String` that results from the `payload.toString()` call). Probably the bytes are extracted again inside `IOUtils.toInputStream` and stored somewhere else while the reference to the temporary `String` is left to be GC'ed.
I think this is almost like forcing it to do `responseStreamingMode="NEVER"` with the exception that it is disguised as a stream and it will require to be sent in chunks. But we are forcing the app to store all the payload in memory, which may not be possible, not to mention the possible performance implications of doing all this pre-loading.

I think you need to continue investigating why this is working. I'm guessing you are getting close. Maybe the `InputStreamHttpEntity` keeps a reference to the payload and that payload is also leaked to the other thread which modifies it.